rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+struct BlockWithTriggers<T: Block> {
+    block: Box<T>,
+    triggers: Vec<dyn NetworkTrigger>,
+}
+
+/// Events emitted by a block stream.
+enum BlockStreamEvent<T: Block> {
+    ChainHead(BlockPointer),
+    Revert(BlockPointer),
+    Block(BlockWithTriggers<T>)
+}
+
+/// Common trait for all block streams.
+trait BlockStream<T: Block>: EventProducer<BlockStreamEvent<T>> {}
+```
+
+Any chain integration has to implement the above traits, except for `NetworkIndexer`
+and `BlockStream`, for which there are default implementations in `graph-core`
+that work with any `Network` implementation.
+
+In addition, each chain will have to be configured in the main `graph-node`
+executable. This RFC does not include details on how this takes place, because
+the configuration options may vary from chain to chain. Ultimately, what
+`graph-node` builds before starting up is a set of supported chains, each with a
+set of supported networks. These are then passed on to, for instance, the
+subgraph instance managemer.
+
+#### Graph CLI
+
+Each chain support comes in the form of a new JS module in the `graph-cli` repo
+under `src/chains/`. The existing Ethereum support is moved to
+`src/chains/ethereum/`.
+
+Each support module is required to export the following structure from its
+`index.js`:
+
+```js
+module.exports = {
+  // Extensions to the GraphQL schema use to validate the manifest.
+  // Could come in the form of new data source types:
+  //
+  //   extend union DataSource SubstrateDataSource
+  //   extend union DataSourceTemplate SubstrateDataSourceTemplate
+  //   ...
+  //
+  manifestExtensions: graphql.Document,
+  
+  // Type conversions to and from AssemblyScript.
+  // Conversions are defined as an array of tuples, each of which
+  // defines the source type (as a pattern or string), the target
+  // type (as a pattern or string), and a code snippet that performs
+  // the conversion.
+  //
+  // An example for Ethereum could look as follows:
+  typeConversions: {
+    fromAssemblyScript: [
+      ['Address', 'address', code => `EthereumValue.fromAddress(${code})`],
+      ['boolean', 'bool', code => `EthereumValue.fromBoolean(${code})`],
+      ['Bytes', 'byte', code => `EthereumValue.fromFixedBytes(${code})`],
+      ['Bytes', 'bytes', code => `EthereumValue.fromBytes(${code})`],
+      [
+        'Bytes',
+        /^bytes(1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32)$/,
+        code => `EthereumValue.fromFixedBytes(${code})`,
+      ],
+      ...
+    ],
+    toAssemblyScript: [
+      ['address', 'Address', code => `${code}.toAddress()`],
+      ['bool', 'boolean', code => `${code}.toBoolean()`],
+      ['byte', 'Bytes', code => `${code}.toBytes()`],
+      [
+        /^bytes(1|2|3|4|5|6|7|8|9|10|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32)?$/,
+        'Bytes',
+        code => `${code}.toBytes()`,
+      ],
+      ...
+    ]
+  },
+  
+  // Code generation for ABIs,
```

Comment:
We won't be able to unify parsing of APIs or a single interface for APIs, but we'll need a way to load ABIs anyway and a way to validate data sources against ABIs.